README.rst
.. -*- coding: utf-8 -*-
.. :Project:   metapensiero.sqlalchemy.asyncpg -- SQLAlchemy adaptor for asyncpg
.. :Created:   Tue 20 Dec 2016 21:17:12 CET
.. :Author:    Lele Gaifax <lele@metapensiero.it>
.. :License:   GNU General Public License version 3 or later
.. :Copyright: © 2016, 2017 Lele Gaifax
..

=================================
 metapensiero.sqlalchemy.asyncpg
=================================

SQLAlchemy adaptor for asyncpg
==============================

 :author: Lele Gaifax
 :contact: lele@metapensiero.it
 :license: GNU General Public License version 3 or later

This is a Python 3 package, spin-off from the proprietary ``Ytefas`` application, that
implements the ability of executing SQLAlchemy core statements through asyncpg__, in a
cleaner way than asyncpgsa__ (I'm biased, of course 😜): the main difference is that it
uses explicitly typed parameter placeholders, to avoid `the problem`__ that initially
prompted me to write this package.

It emits ``DEBUG`` logs with `prettified`__ SQL statements, with *parameters* resolved to
the actual *values*, and supplies an asyncpg variant of `metapensiero.sqlalchemy.proxy`__
\ 's ProxiedQuery__.


__ https://pypi.python.org/pypi/asyncpg
__ https://pypi.python.org/pypi/asyncpgsa
__ https://github.com/MagicStack/asyncpg/issues/32
__ https://pypi.python.org/pypi/metapensiero.sqlalchemy.proxy
__ http://metapensierosqlalchemyproxy.readthedocs.io/en/latest/\
   core.html#metapensiero.sqlalchemy.proxy.core.ProxiedQuery
__ http://pg-query.readthedocs.io/en/latest/


Typical usage
-------------

The following script:

.. code-block:: python

  import asyncio
  from datetime import date
  import logging

  import sqlalchemy as sa
  from asyncpg import create_pool
  from asyncpg.types import Range
  from metapensiero.sqlalchemy import asyncpg as sasyncpg


  async def dml_tests(connection):
      sasyncpg_test = sa.Table('sasyncpg_test', sa.MetaData(),
                               sa.Column('id', sa.types.Integer, primary_key=True),
                               sa.Column('value', sa.types.Text),
                               sa.Column('period', sa.dialects.postgresql.DATERANGE))

      value = 'First test'
      insert_stmt = sasyncpg_test.insert().values(id=1, value=value)
      await connection.execute(insert_stmt)

      new_value = 'Second test'
      update_stmt = (sasyncpg_test.update()
                     .values(value=new_value,
                             period=Range(date(2016, 2, 1), date(2016, 3, 1)))
                     .where(sasyncpg_test.c.id == 1))
      await connection.execute(update_stmt)

      select_stmt = (sa.select([sasyncpg_test.c.value])
                     .where(sasyncpg_test.c.id == sa.bindparam('id')))
      for row in await connection.fetchall(select_stmt, named_args={'id': 1}):
          print('Row:', row)

      single_row = (sa.select([sasyncpg_test])
                    .where(sasyncpg_test.c.period.contains(date(2016, 2, 15))))
      print('Row:', await connection.fetchone(single_row))


  async def run(loop):
      pool = await create_pool(database="test", loop=loop)

      async with pool.acquire() as apgc:
          connection = sasyncpg.Connection(apgc)

          query = sa.select([sa.func.version()])
          result = await connection.scalar(query)
          print("PostgreSQL version:", result)

          await connection.execute('create table sasyncpg_test ('
                                   ' id integer not null primary key,'
                                   ' value text,'
                                   ' period daterange)')

          try:
              await dml_tests(connection)
          finally:
              await connection.execute('DROP TABLE sasyncpg_test')


  def main():
      loop = asyncio.get_event_loop()
      loop.run_until_complete(run(loop))


  if __name__ == '__main__':
      logging.basicConfig(level=logging.DEBUG)
      main()

produces something like::

  DEBUG:asyncio:Using selector: EpollSelector
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Fetching scalar in transaction 9ddb60:
      SELECT version() AS version_1
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Fetched value in 869 µsec
  PostgreSQL version: PostgreSQL 9.6.6 on x86_64-pc-linux-gnu, compiled by gcc (Debian 7.2.0-12) 7.2.1 20171025, 64-bit
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Executing in transaction 9ddb60:
      CREATE TABLE sasyncpg_test (
          id integer NOT NULL PRIMARY KEY, value text, period daterange
      )
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Execution took 97.1 msec
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Executing in transaction 9ddb60:
      INSERT INTO sasyncpg_test (id, value)
      VALUES (1::integer, 'First test'::text)
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Execution took 1.26 msec
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Executing in transaction 9ddb60:
      UPDATE sasyncpg_test
      SET value = 'Second test'::text, period = '[2016-02-01,2016-03-01)'::daterange
      WHERE sasyncpg_test.id = 1::integer
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Execution took 14.1 msec
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Fetching rows in transaction 9ddb60:
      SELECT sasyncpg_test.value
      FROM sasyncpg_test
      WHERE sasyncpg_test.id = 1::integer
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Fetched 1 records in 909 µsec
  Row: <Record value='Second test'>
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Fetching row in transaction 9ddb60:
      SELECT sasyncpg_test.id, sasyncpg_test.value, sasyncpg_test.period
      FROM sasyncpg_test
      WHERE sasyncpg_test.period @> datetime.date(2016, 2, 15)::date
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Fetched one record in 951 µsec
  Row: <Record id=1 value='Second test' period=<Range [datetime.date(2016, 2, 1), datetime.date(2016, 3, 1))>>
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Executing in transaction 9ddb60:
      DROP TABLE sasyncpg_test RESTRICT
  DEBUG:metapensiero.sqlalchemy.asyncpg.funcs:Execution took 12 msec


CHANGES.rst
.. -*- coding: utf-8 -*-

Changes
-------

0.1 (2017-12-03)
~~~~~~~~~~~~~~~~

- Standalone package, released under GPLv3


0.0 (unreleased)
~~~~~~~~~~~~~~~~

- Initial effort.


src/metapensiero/sqlalchemy/asyncpg/funcs.py
# -*- coding: utf-8 -*-
# :Project:   metapensiero.sqlalchemy.asyncpg -- Adaptor functions
# :Created:   Tue 20 Dec 2016 21:17:12 CET
# :Author:    Lele Gaifax <lele@metapensiero.it>
# :License:   GNU General Public License version 3 or later
# :Copyright: © 2016, 2017 Lele Gaifax
#

import logging
from time import perf_counter

from .dialect import PGDialect_asyncpg


SLOW_QUERY_THRESHOLD = 2.0
"Warn about SQL statements that take more than this amount of seconds."

logger = logging.getLogger(__name__)


def _honor_column_default(params, column, default, key_getter):
    key = key_getter(column)
    val = params.get(key)
    if val is None:
        if not default.is_sequence and default.is_scalar:
            val = default.arg
        else:
            if default.is_callable:
                val = default.arg(None)

        if val is not None:
            params[key] = val


def _format_arg(arg):
    from uuid import UUID
    from asyncpg.types import Range

    if isinstance(arg, UUID):
        arg = str(arg)
    elif isinstance(arg, Range):
        if arg.isempty:
            arg = 'empty range'
        else:
            lb = '[' if arg.lower_inc else '('
            ub = ']' if arg.upper_inc else ')'
            v = arg.lower
            lv = '' if v is None else v.isoformat()
            v = arg.upper
            uv = '' if v is None else v.isoformat()
            arg = '%s%s,%s%s' % (lb, lv, uv, ub)

    rarg = repr(arg)

    if len(rarg) > 50:
        rarg = rarg[:20] + ' … ' + rarg[-20:]

    return rarg


def _format_elapsed_time(et):
    precision = 3
    units = {"nsec": 1e-9, "µsec": 1e-6, "msec": 1e-3, "sec": 1.0}
    scales = [(scale, unit) for unit, scale in units.items()]
    scales.sort(reverse=True)
    for scale, unit in scales:
        if et >= scale:
            break
    return "%.*g %s" % (precision, et / scale, unit)


def _log_sql_statement(connection, operation, sql, args, logf=logger.debug):
    from textwrap import indent
    from asyncpg.pool import PoolConnectionProxy
    from pg_query import prettify
    from pg_query.printer import get_printer_for_node_tag, node_printer
    import pg_query.printers.dml  # noqa

    try:
        orig_paramref_printer = get_printer_for_node_tag(None, 'ParamRef')
        try:
            @node_printer('ParamRef', override=True)
            def replace_param_ref(node, output):
                if 0 < node.number.value <= len(args):
                    output.write(_format_arg(args[node.number.value - 1]))
                else:
                    orig_paramref_printer(node, output)
            newsql = prettify(sql, compact_lists_margin=80, safety_belt=False)
        finally:
            node_printer('ParamRef', override=True)(orig_paramref_printer)
    except Exception as e:
        logger.error('Something wrong with SQL prettification: %s\n'
                     '  arguments: %r', e, [_format_arg(a) for a in args])
    else:
        sql = newsql

    sql = indent(sql, '    ')

    if isinstance(connection, PoolConnectionProxy):
        tx = connection._con._top_xact
    else:
        tx = connection._top_xact

    logf('%s in transaction %0x:\n%s', operation, id(tx), sql)


def compile(stmt, pos_args=None, named_args=None, _d=PGDialect_asyncpg()):
    """Compile an SQLAlchemy core statement and extract its parameters.

    :param stmt: any SQLAlchemy core statement or a raw SQL instruction
    :param pos_args: a possibly empty sequence of positional arguments
    :param named_args: a possibly empty mapping of named arguments
    :return: a tuple of two items, a string with the final SQL instruction and
             a sequence of positional arguments, if any

    If `stmt` is a plain string, it is returned as is together with a tuple of
    the positional arguments specified in `pos_args`, possibly empty.

    When `stmt` is a SQLAlchemy :class:`Executable
    <sqlalchemy.sql.base.Executable>` it is compiled to its raw SQL using a
    custom SA :class:`dialect <.PGDialect_asyncpg>` to produce a syntax
    compatible with asyncpg (in particular, using ``$1``, ``$2`` and so on as
    the parametric placeholders) and it's parameters collected in a tuple,
    picking named parameters (when using :func:`sqlalchemy.bindparam` for
    example) from the `named_args` dictionary.

    The result is suitable to be passed to the various methods of asyncpg's
    connection for execution.
    """

    if isinstance(stmt, str):
        return stmt, tuple(pos_args) if pos_args is not None else ()
    else:
        compiled = stmt.compile(dialect=_d)
        params = compiled.construct_params(named_args)

        # Honor column's default or unupdate setting: following code adapted
        # from SA's DefaultExecutionContext._process_executesingle_defaults()
        # logic

        key_getter = compiled._key_getters_for_crud_column[2]
        for c in compiled.insert_prefetch:
            default = c.default
            if default is not None:
                _honor_column_default(params, c, default, key_getter)

        for c in compiled.update_prefetch:
            default = c.onupdate
            if default is not None:
                _honor_column_default(params, c, default, key_getter)

        return compiled.string, tuple(params[p] for p in compiled.positiontup)


class UnexpectedResultError(RuntimeError):
    "Exception raised when the execution result does not match."

    def __init__(self, got, expected):
        super().__init__(f'Expected {expected!r}, got {got!r}')
        self.expected = expected
        self.got = got


async def execute(apgconn, stmt, pos_args=None, named_args=None,
                  expected_result=None,
                  warn_slow_query_threshold=SLOW_QUERY_THRESHOLD, **kwargs):
    r"""Execute the given statement on a asyncpg connection.

    :param apgconn: an ASyncPG Connection__ instance
    :param stmt: any SQLAlchemy core statement or a raw SQL instruction
    :param pos_args: a possibly empty sequence of positional arguments
    :param named_args: a possibly empty mapping of named arguments
    :param expected_result: the expected result of the execution, if any
    :param \*\*kwargs: any valid `execute()`__ keyword argument
    :return: a string with the status of the last instruction

    The `stmt` is first compiled with :func:`.compile` and then executed on
    the `apgconn` connection with the needed parameters.

    If `expected_result` is not ``None``, then it must match the value
    returned by the underlying function, otherwise an
    :class:`UnexpectedResultError` is raised.

    __ https://magicstack.github.io/asyncpg/devel/api/index.html#connection
    __ https://magicstack.github.io/asyncpg/devel/api/\
       index.html#asyncpg.connection.Connection.execute
    """

    sql, args = compile(stmt, pos_args, named_args)
    debug = logger.isEnabledFor(logging.DEBUG)
    if debug:
        _log_sql_statement(apgconn, 'Executing', sql, args)
    t0 = perf_counter()
    try:
        result = await apgconn.execute(sql, *args, **kwargs)
    except Exception as e:
        if not debug:
            _log_sql_statement(apgconn, 'Error "%s" executing' % e,
                               sql, args, logf=logger.error)
        raise
    else:
        if expected_result is not None and result != expected_result:
            if not debug:
                _log_sql_statement(apgconn, 'Unexpected result executing',
                                   sql, args, logf=logger.error)
            raise UnexpectedResultError(result, expected_result)

    elapsed = perf_counter() - t0
    if debug or elapsed > warn_slow_query_threshold:
        if debug:
            logf = (logger.debug if elapsed < warn_slow_query_threshold
                    else logger.warning)
            logf('Execution took %s', _format_elapsed_time(elapsed))
        else:
            _log_sql_statement(apgconn, 'Suspiciously SLOW query', sql, args,
                               logf=logger.warning)
    return result


async def prepare(apgconn, stmt, **kwargs):
    r"""Create a `prepared statement`__.

    :param apgconn: an asyncpg Connection__ instance
    :param stmt: any SQLAlchemy core statement or a raw SQL instruction
    :param \*\*kwargs: any valid `prepare()`__ keyword argument
    :return: a string with the status of the last instruction

    __ https://magicstack.github.io/asyncpg/devel/api/\
       index.html#prepared-statements
    __ https://magicstack.github.io/asyncpg/devel/api/index.html#connection
    __ https://magicstack.github.io/asyncpg/devel/api/\
       index.html#asyncpg.connection.Connection.prepare
    """

    sql, args = compile(stmt)
    debug = logger.isEnabledFor(logging.DEBUG)
    if debug:
        _log_sql_statement(apgconn, 'Preparing', sql, args)
        t0 = perf_counter()
    try:
        result = await apgconn.prepare(sql, **kwargs)
    except Exception as e:
        if not debug:
            _log_sql_statement(apgconn, 'Error "%s" preparing' % e,
                               sql, args, logf=logger.error)
        raise
    if debug:
        t1 = perf_counter()
        logger.debug('Preparation took %s', _format_elapsed_time(t1 - t0))
    return result


async def fetchall(apgconn, stmt, pos_args=None, named_args=None,
                   warn_slow_query_threshold=SLOW_QUERY_THRESHOLD, **kwargs):
    r"""Execute the given statement on a asyncpg connection and return
    resulting records.

    :param apgconn: an asyncpg Connection__ instance
    :param stmt: any SQLAlchemy core statement or a raw SQL instruction
    :param pos_args: a possibly empty sequence of positional arguments
    :param named_args: a possibly empty mapping of named arguments
    :param \*\*kwargs: any valid `fetch()`__ keyword argument
    :return: a list of `Record`__ instances

    The `stmt` is first compiled with :func:`.compile` and then executed on
    the `apgconn` connection with the needed parameters. It's whole resultset
    is finally returned.

    __ https://magicstack.github.io/asyncpg/devel/api/index.html#connection
    __ https://magicstack.github.io/asyncpg/devel/api/\
       index.html#asyncpg.connection.Connection.fetch
    __ https://magicstack.github.io/asyncpg/devel/api/\
       index.html#asyncpg.Record
    """

    sql, args = compile(stmt, pos_args, named_args)
    debug = logger.isEnabledFor(logging.DEBUG)
    if debug:
        _log_sql_statement(apgconn, 'Fetching rows', sql, args)
    t0 = perf_counter()
    try:
        result = await apgconn.fetch(sql, *args, **kwargs)
    except Exception as e:
        if not debug:
            _log_sql_statement(apgconn, 'Error "%s" fetching rows' % e,
                               sql, args, logf=logger.error)
        raise
    elapsed = perf_counter() - t0
    if debug or elapsed > warn_slow_query_threshold:
        if debug:
            logf = (logger.debug if elapsed < warn_slow_query_threshold
                    else logger.warning)
            logf('Fetched %d records in %s',
                 len(result), _format_elapsed_time(elapsed))
        else:
            _log_sql_statement(apgconn, 'Suspiciously SLOW query', sql, args,
                               logf=logger.warning)
    return result


async def fetchone(apgconn, stmt, pos_args=None, named_args=None,
                   warn_slow_query_threshold=SLOW_QUERY_THRESHOLD, **kwargs):
    r"""Execute the given statement on a asyncpg connection and return the
    first row.

    :param apgconn: an asyncpg Connection__ instance
    :param stmt: any SQLAlchemy core statement or a raw SQL instruction
    :param pos_args: a possibly empty sequence of positional arguments
    :param named_args: a possibly empty mapping of named arguments
    :param \*\*kwargs: any valid `fetchrow()`__ keyword argument
    :return: either ``None`` or a `Record`__ instance

    The `stmt` is first compiled with :func:`.compile` and then executed on
    the `apgconn` connection with the needed parameters. The first row of it's
    result is finally returned, or ``None`` if the result is empty.

    __ https://magicstack.github.io/asyncpg/devel/api/index.html#connection
    __ https://magicstack.github.io/asyncpg/devel/api/\
       index.html#asyncpg.connection.Connection.fetchrow
    __ https://magicstack.github.io/asyncpg/devel/api/\
       index.html#asyncpg.Record
    """

    sql, args = compile(stmt, pos_args, named_args)
    debug = logger.isEnabledFor(logging.DEBUG)
    if debug:
        _log_sql_statement(apgconn, 'Fetching row', sql, args)
    t0 = perf_counter()
    try:
        result = await apgconn.fetchrow(sql, *args, **kwargs)
    except Exception as e:
        if not debug:
            _log_sql_statement(apgconn, 'Error "%s" fetching row' % e,
                               sql, args, logf=logger.error)
        raise
    elapsed = perf_counter() - t0
    if debug or elapsed > warn_slow_query_threshold:
        if debug:
            logf = (logger.debug if elapsed < warn_slow_query_threshold
                    else logger.warning)
            logf('Fetched %s in %s',
                 'no records' if result is None else 'one record',
                 _format_elapsed_time(elapsed))
        else:
            _log_sql_statement(apgconn, 'Suspiciously SLOW query', sql, args,
                               logf=logger.warning)
    return result


async def scalar(apgconn, stmt, pos_args=None, named_args=None,
                 warn_slow_query_threshold=SLOW_QUERY_THRESHOLD, **kwargs):
    r"""Execute the given statement on a asyncpg connection and return a
    single column of the first row.

    :param apgconn: an asyncpg Connection__ instance
    :param stmt: any SQLAlchemy core statement or a raw SQL instruction
    :param pos_args: a possibly empty sequence of positional arguments
    :param named_args: a possibly empty mapping of named arguments
    :param \*\*kwargs: any valid `fetchval()`__ keyword argument
    :return: the value of the specified column of the first record, or
             ``None`` if the query does not return any rows

    The `stmt` is first compiled with :func:`.compile` and then executed on
    the `apgconn` connection with the needed parameters. A particular column
    of the first row (specified positionally with the keyword `column`, by
    default ``0`` to mean the first) is finally returned.

    .. warning:: There is no way to distinguish between a *non existing
                 record* and a ``None`` value in the requested column of the
                 first row. If you need to discriminate the case, use
                 :meth:`fetchone()` instead:

                 .. code-block:: python

                    rec = conn.fetchone(stmt, pos_args)
                    if rec is not None:
                        val = rec[column]
                        ...

    __ https://magicstack.github.io/asyncpg/devel/api/index.html#connection
    __ https://magicstack.github.io/asyncpg/devel/api/\
       index.html#asyncpg.connection.Connection.fetchval
    """

    sql, args = compile(stmt, pos_args, named_args)
    debug = logger.isEnabledFor(logging.DEBUG)
    if debug:
        _log_sql_statement(apgconn, 'Fetching scalar', sql, args)
    t0 = perf_counter()
    try:
        result = await apgconn.fetchval(sql, *args, **kwargs)
    except Exception as e:
        if not debug:
            _log_sql_statement(apgconn, 'Error "%s" fetching scalar' % e,
                               sql, args, logf=logger.error)
        raise
    elapsed = perf_counter() - t0
    if debug or elapsed > warn_slow_query_threshold:
        if debug:
            logf = (logger.debug if elapsed < warn_slow_query_threshold
                    else logger.warning)
            logf('Fetched value in %s', _format_elapsed_time(elapsed))
        else:
            _log_sql_statement(apgconn, 'Suspiciously SLOW query', sql, args,
                               logf=logger.warning)
    return result
